crates/wasmi/src/memory/mod.rs

```diff
@@ -165,6 +165,49 @@ impl MemoryEntity {
         }
     }
 
+    /// Creates a new memory entity with the given memory type.
+    pub fn new_static(
```

Comment:
The main design question I have is the following: Shall it be possible for a static memory based `Memory` instance to be growable? Technically this does not make sense since the buffer is static anyways. Also for certain use cases such as when fuel metering is enabled the advantage of ungrowable static memory is that `memory.grow` would cost O(1) fuel guaranteed.
This requires `initial_pages == current_pages == static_buffer.len() * BYTES_PER_PAGE` always for static memory based `Memory` instances.

Is there a use case that would suffer from this design decision?

With the current design the static memory buffer already is _not_ growable (obviously), however `Memory::grow` still performs (hopeful) work to do so.

Reply:
In case we want to change growability of static memory based `Memory` instances we should reflect that in the documentation of the `new_static` constructor.

Reply:
Some WASM allocators are requesting pages (minimum 1), but I'm not sure would it get one if memory is already at max capacity, do you know? Maybe if memory would be allowed to grow, they can get their page. That is the only possible problem I can think of, I don't know is it true though

Reply:
The `memory.grow` instruction cannot grow linear memory if it is already at its maximum size. However, failing to grow linear memory is something that should be expected hence failure won't trap but return a magic value which can (and should) be handled by a memory allocator. I am not sure how much memory allocators rely on growing linear memory but it would be very good to know for sure.

Reply:
I found an answer to our problem. For static memories we need to zero all the bytes of the new pages whenever we grow it. This is done implicitly with the non-static buffer backend (via `Vec`) but needs to be done with the static buffer upon grow because we are not guaranteed that the provided static buffer is already zeroed. And then we resolved this question because now it makes sense to do linear work for growing memories backed by a static buffer.

Reply:
This needs to be done for both, growing memory and when creating new linear memory from a static buffer.
When growing memory bytes for the new pages need to be zeroed and when creating a new linear memory from a static buffer the initial pages need to be zeroed.

Reply:
More info: https://webassembly.github.io/spec/core/exec/modules.html#growing-memories

Reply:
If wasmi will zero those bytes, maybe request `&'static mut [MaybeUninit<u8>]`?

Reply:
I guess keeping everything safe and simple would be better, because zeroing would probably be very fast and quite infrequent, but it will make *some* difference because static memory is not required to be initialized.

Reply:
Okay then please go ahead and implement the changes with `&'mut static [u8]` and we can always change it later to `&'static mut [MaybeUninit<u8>]` if we see a good enough reason to do so. :)